### PR TITLE
Add covariance matrix function theano.tensor.cov

### PR DESCRIPTION
**Pull request by ieee8023**
I added a simple function theano.tensor.cov using existing theano functions so that it may run on a GPU. This is is modeled after np.cov but matches only one of it's function signatures.

These guys helped me figure this out!
Olexa Bilaniuk @obilaniu
Zhouhan Lin @hantek

### CODE REVIEW

**Comment by JesseLivezey** [comment]
It's probably be a good idea to make the interface more similar to the ```numpy.cov``` function
https://docs.scipy.org/doc/numpy/reference/generated/numpy.cov.html

```
cov(m, rowvar=True, bias=False, ddof=None)
```
would probably be a reasonable minimal interface.

**Comment by ieee8023** [comment]
@JesseLivezey I agree it will be more useful to have this functionality. I can get to this next week! Thanks for the feedback.

**Comment by lamblin** [comment]
Thanks! I agree that mimicking the full functionality of NumPy would be better.
Also, the automatic flake8 test currently [fails with these errors](https://travis-ci.org/Theano/Theano/jobs/219138467#L893), which would probably be easy to fix.

**Comment by JesseLivezey** [comment]
The full ```numpy.cov``` signature is a little more complex than what I posted above
```
numpy.cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None, aweights=None)
```
but they list the full computation on this page
https://docs.scipy.org/doc/numpy/reference/generated/numpy.cov.html
and you can look at the numpy code here
https://github.com/numpy/numpy/blob/561f1accf861ad8606ea2dd723d2be2b09a2dffa/numpy/lib/function_base.py#L2725
so it might not be that much more work to write something a little more general.

**Comment by nouiz** [comment]
Just a comment. There is no problem to implement only a subset of np.cov, but having everything is better as other said :)

But if you don't implement everything, make sure to have a fct signature compatible with np.cov. For example, if you don't implement y, you still need to keep it in the signature and raise not implemented if someone try to use it. As if we add support for it later, we don't want to break the old interface by changing the order of parameter.

thanks for the contribution.

**Comment by ieee8023** [comment]
I implemented all the arguments mentioned initially by @JesseLivezey. The weight arguments were a bit too complicated so I made the function raise a NotImplementedError if they are set as @nouiz said.

I added a bunch of tests. If I missed a use case let me know and I can work more on it.

File: theano/tensor/basic.py [hunk 1]
```diff
@@ -2168,6 +2168,67 @@ def sqr(a):
...
 
+def cov(X, y=None, rowvar=True, bias=False, ddof=None, fweights=None, aweights=None):
+    """Calculate the covariance matrix.
+    Covariance indicates the level to which two variables vary together.
+    If we examine N-dimensional samples, :math:`X = [x_1, x_2, ... x_N]^T`,
+    then the covariance matrix element :math:`C_{ij}` is the covariance of
+    :math:`x_i` and :math:`x_j`. The element :math:`C_{ii}` is the variance
+    of :math:`x_i`. Code and docstring ported from numpy.
+    ----------
+    a : array_like
```
Comment (JesseLivezey) [inline theano/tensor/basic.py]: I would change ```X``` above and ```a``` here to ```m``` to be consistent with numpy.cov.

Comment (ieee8023) [inline theano/tensor/basic.py]: Maybe it should all be `a` to match the other theano code?

File: theano/tensor/tests/test_basic.py [hunk 2]
```diff
@@ -8172,6 +8172,107 @@ def test_norm():
...
 
+class test_cov(unittest.TestCase):
+
+    def test_core(self):
+        x = theano.tensor.matrix('x')
+        c = theano.tensor.cov(x)
+        f = theano.function([x], c)
+
+        # basic cov function
+        data = np.asarray(np.random.rand(3, 5), dtype=config.floatX)
```
Comment (JesseLivezey) [inline theano/tensor/tests/test_basic.py]: This series of test could be written more concisely as a loop over shapes. Same for other tests.

File: theano/tensor/basic.py (hunk 1 above)
Comment (abergeron) [inline theano/tensor/basic.py]: We try to mimic the numpy API in general so going with what numpy does is not a bad idea.  In any case, this absolutely needs to have the same name in the function definition and the doc.

**Comment by nouiz** [comment]
We try to make Theano 0.10. It would be great that this is finished and included. We don't have time to finish it ourself, so if someone have time to finish it this week, it would be great!

**Comment by JesseLivezey** [comment]
@ieee8023 I think the only thing you need to do is change the variable names in the code and documentation to be consistent with numpy.

**Comment by ieee8023** [comment]
Ok, I can get this done before the end of the week

**Pull request merged**